manifests/config.pp

```diff
@@ -34,139 +34,160 @@
-  $pool_remove_abandoned = pick($jira::pool_remove_abandoned, true)
-  $pool_remove_abandoned_timeout = pick($jira::pool_remove_abandoned_timeout, 300)
-  $min_evictable_idle_time = pick($jira::min_evictable_idle_time, 60000)
-  $time_between_eviction_runs = pick($jira::time_between_eviction_runs, 300000)
-  $pool_test_while_idle = pick($jira::pool_test_while_idle, true)
-  $pool_test_on_borrow = pick($jira::pool_test_on_borrow, false)
-
-  # JIRA will complain if these aren't set for PostgreSQL (will work fine though)
-  # https://confluence.atlassian.com/jirakb/connection-problems-to-postgresql-result-in-stuck-threads-in-jira-1047534091.html
-  if $jira::db == 'postgresql' {
-    $connection_settings = pick($jira::connection_settings, 'tcpKeepAlive=true;socketTimeout=240')
-  } else {
-    $connection_settings = $jira::connection_settings
-  }
-
-  if $jira::db == 'mysql' {
-    $validation_query_timeout = pick($jira::validation_query_timeout, 3)
+  $validation_query = if $jira::validation_query {
+    $jira::validation_query
   } else {
-    $validation_query_timeout = $jira::validation_query_timeout
-  }
-
-  if $jira::validation_query {
-    $validation_query = $jira::validation_query
-  } else {
-    $validation_query = $jira::db ? {
+    $jira::db ? {
       'mysql'      => 'select 1',
       'sqlserver'  => 'select 1',
       'oracle'     => 'select 1 from dual',
       'postgresql' => 'select version();',
       'h2'         => undef,
     }
   }
-
-  $tomcat_protocol_ssl_real = pick($jira::tomcat_protocol_ssl, 'org.apache.coyote.http11.Http11NioProtocol')
-
-  $jira_properties = {
-    'jira.websudo.is.disabled' => !$jira::enable_secure_admin_sessions,
-  }
-  $merged_jira_config_properties = jira::sort_hash($jira_properties + $jira::jira_config_properties)
-
-  # Configuration logic ends, resources begin:
-
-  file { "${jira::webappdir}/bin/user.sh":
-    content => epp("${module_name}/user.sh.epp"),
-    mode    => '0755',
+  $validation_query_timeout = if $jira::validation_query_timeout {
+    $jira::validation_query_timeout
+  } else {
+    $jira::db ? {
+      'mysql' => 3,
+      default => undef,
+    }
   }
-
-  file { "${jira::webappdir}/bin/setenv.sh":
-    content => epp("${module_name}/setenv.sh.epp"),
-    mode    => '0755',
+  # JIRA will complain if these aren't set for PostgreSQL (will work fine though)
+  # https://confluence.atlassian.com/jirakb/connection-problems-to-postgresql-result-in-stuck-threads-in-jira-1047534091.html
+  $connection_settings = if $jira::connection_settings {
+    $jira::connection_settings
+  } else {
+    $jira::db ? {
+      'postgresql' => 'tcpKeepAlive=true;socketTimeout=240',
+      default => undef,
+    }
   }
-
-  $dbconfig_template = $jira::use_jndi_ds ? {
-    true    => "${module_name}/dbconfig.jndi.xml.epp",
-    default => "${module_name}/dbconfig.xml.epp"
+  $dbconfig_params = {
+    version => $jira::version,
+    is_installed => Deferred('jira::is_installed', [$jira::homedir]),
```

Comment:
When does this get run?  Could be an issue depending on this setting???https://www.puppet.com/docs/puppet/8/configuration.html#preprocess-deferred

Reply:
Good question :)
I just followed the example from docs https://www.puppet.com/docs/puppet/8/deferring_functions.html#deferring_functions-deferred-function-example

Reply:
Ran a test with this in puppet.conf on the first puppet run on the jira node
```
[agent]
    preprocess_deferred = true
```
No issues detected.
I guess plugin retrieval is done even before preprosessing deferrals.